src/purge.rs

```diff
@@ -41,6 +51,68 @@ where
     // This table records Raft Groups that should be force compacted before. Those that are not
     // compacted in time (after `MAX_EPOCH_BEFORE_FORCE_REWRITE` epochs) will be force rewritten.
     force_rewrite_candidates: Arc<Mutex<HashMap<u64, u32>>>,
+
+    // The rewrite of the entire rewrite queue of a Raft Group should be atomic.
+    // We use a special marker to indicate such a rewrite is on-going. Before recovery, rewrite
+    // files that are newer than this marker (have bigger seqno) will be removed.
+    rewrite_marker_path: PathBuf,
+}
+
+fn write_marker(p: &Path, seq: u64) -> Result<()> {
+    use std::io::Write;
+    let mut f = fs::File::create(p)?;
+    f.write_all(seq.to_string().as_bytes())?;
+    f.sync_all()?;
+    f.write_all(b"//")?;
+    f.sync_all()?;
+    if let Some(dir) = p.parent() {
+        fs::File::open(dir).and_then(|d| d.sync_all())?;
+    }
+    Ok(())
+}
+
+fn read_marker(p: &Path) -> Result<Option<u64>> {
+    match fs::read_to_string(p) {
+        Err(e) if e.kind() != std::io::ErrorKind::NotFound => Err(e.into()),
+        Ok(s) if s.ends_with("//") => Ok(Some(s[..s.len() - 2].parse::<u64>().unwrap())),
+        _ => Ok(None),
+    }
+}
+
+fn remove_marker(p: &Path) -> Result<()> {
+    fs::remove_file(p)?;
+    if let Some(dir) = p.parent() {
+        fs::File::open(dir).and_then(|d| d.sync_all())?;
+    }
+    Ok(())
+}
+
+pub fn before_recover(cfg: &Config, fs: &impl crate::env::FileSystem) -> Result<()> {
+    use crate::file_pipe_log::FileNameExt;
+    let mut rewrite_marker_path = PathBuf::from(&cfg.dir);
+    rewrite_marker_path.push(REWRITE_MARKER);
+    if let Some(seq) = read_marker(&rewrite_marker_path)? {
+        let mut f = FileId::new(LogQueue::Rewrite, seq + 1);
+        let mut paths = Vec::new();
+        loop {
+            let path = f.build_file_path(&cfg.dir);
```

Comment:
What if new appends are flushed into a new file before purge_expires_files finished?

Reply:
This is called before opening engine, no one can write.

Reply:
But the file is created during rewrite, right? And rewrite can happen anytime?

Reply:
I still don't understand your case.